### DISCUSSION TITLE
How to resolve Union types in pd.concat?

### DISCUSSION THREAD

**Discussion by tvdboom** [pandas-dev/pandas-stubs#878] [Q&A]
Hi,

When running this code with `mypy` (note I am using the code in `main`, which accepts `None` as type for pd.concat):

```python
data: pd.DataFrame | None = pd.DataFrame()

pd.concat([pd.DataFrame(), data])
```

it returns this error:

`error: List item 1 has incompatible type "DataFrame | None"; expected "DataFrame"  [list-item]`

Now, I understand where the error comes from, the overloaded method has type `Iterable[DataFrame | None]` and `data` has type `Union[DataFrame, None]` instead of only `DataFrame` or `None`, hence it's incompatible. But common sense says this should work, and indeed, at runtime the code works just fine. Is there a clean/standard way of solving this error on `mypy`?

I understand I could just write an if-else statement to check if data is None, but I am looking for a more structural solution that could solve other similar issues, for example when `data` can also be a `pd.Series`, e.g., `data: pd.Series | pd.DataFrame | None`.

Thank you.

**Comment by Dr-Irv** [comment]
I just tried your example on what is currently in `main` and it works fine with `mypy` and `pyright`.

**Comment by tvdboom** [comment]
You are right, the example doesn't work, my bad. This example does work though, and the same question still applies.

```python
data: pd.DataFrame | pd.Series = pd.Series()

pd.concat([pd.DataFrame(), data])
```

returns

```
error: List item 0 has incompatible type "DataFrame"; expected "None"  [list-item]
error: List item 1 has incompatible type "DataFrame | Series[Any]"; expected "None"  [list-item]
```

**Comment by Dr-Irv** [comment]
I created an issue #879
